File: patterns/Creational/Builder/JavaScript/RealWorldExample/MealPlan.js
// Builder Design Pattern - Meal Plan
//
// This example demonstrates the Builder Design Pattern to create different
// types of meal plans (Vegetarian, Standard).

// MealPlanBuilder interface specifies methods for creating different parts of the
// Meal Plan.
class MealPlanBuilder {
    addMainCourse() {}
    addSideDish() {}
    addDessert() {}
    getMealPlan() {}
}

// Concrete Builder class for creating a Standard Meal Plan.
class StandardMealPlanBuilder extends MealPlanBuilder {
    constructor() {
        super();
        this.mealPlan = new MealPlan();
    }

    addMainCourse() {
        this.mealPlan.mainCourse = "Steak";
        return this;
    }

    addSideDish() {
        this.mealPlan.sideDish = "French Fries";
        return this;
    }

    addDessert() {
        this.mealPlan.dessert = "Ice Cream";
        return this;
    }

    getMealPlan() {
        return this.mealPlan;
    }
}

// Concrete Builder class for creating a Vegetarian Meal Plan.
class VegetarianMealPlanBuilder extends MealPlanBuilder {
    constructor() {
        super();
        this.mealPlan = new MealPlan();
    }

    addMainCourse() {
        this.mealPlan.mainCourse = "Vegetarian Burger";
        return this;
    }

    addSideDish() {
        this.mealPlan.sideDish = "Salad";
        return this;
    }

    addDessert() {
        this.mealPlan.dessert = "Fruit Salad";
        return this;
    }

    getMealPlan() {
        return this.mealPlan;
    }
}

// The Product class represents the complex object under construction, in this
// case, a Meal Plan.
class MealPlan {
    constructor() {
        this.mainCourse = "";
        this.sideDish = "";
        this.dessert = "";
    }

    listItems() {
        console.log(`Main Course: ${this.mainCourse}`);
        console.log(`Side Dish: ${this.sideDish}`);
        console.log(`Dessert: ${this.dessert}`);
    }
}

// The Director class defines the order in which the construction steps
// should be called to build the final meal plan.
class MealPlanDirector {
    buildStandardMeal(builder) {
        return builder.addMainCourse()
            .addSideDish()
            .addDessert()
            .getMealPlan();
    }

    buildVegetarianMeal(builder) {
        return builder.addMainCourse()
            .addSideDish()
            .addDessert()
            .getMealPlan();
    }
}

// Client code to demonstrate the Meal Plan creation process.
function clientCode(director, builder) {
    let mealPlan = director.buildStandardMeal(builder);
    console.log("Standard Meal Plan:");
    mealPlan.listItems();

    console.log();

    let vegetarianMealPlan = director.buildVegetarianMeal(builder);
    console.log("Vegetarian Meal Plan:");
    vegetarianMealPlan.listItems();
}

// Usage Example:
const director = new MealPlanDirector();

console.log("Testing Standard Meal Plan Builder:");
clientCode(director, new StandardMealPlanBuilder());

console.log("\n\nTesting Vegetarian Meal Plan Builder:");
clientCode(director, new VegetarianMealPlanBuilder());

File: patterns/Creational/Builder/Java/RealWorldExample/MealPlan.java
package patterns.Creational.Builder.Java.RealWorldExample;

// Builder Design Pattern - Meal Plan
//
// This example demonstrates the Builder Design Pattern to create different
// types of meal plans (Vegetarian, Standard).

// The Builder interface specifies methods for creating different parts of the
// Meal Plan.
interface MealPlanBuilder {
    MealPlanBuilder addMainCourse();
    MealPlanBuilder addSideDish();
    MealPlanBuilder addDessert();
    MealPlan getMealPlan();
}

// Concrete Builder class for creating a Standard Meal Plan.
class StandardMealPlanBuilder implements MealPlanBuilder {
    private MealPlan mealPlan;

    public StandardMealPlanBuilder() {
        this.mealPlan = new MealPlan();
    }

    @Override
    public MealPlanBuilder addMainCourse() {
        mealPlan.setMainCourse("Steak");
        return this;
    }

    @Override
    public MealPlanBuilder addSideDish() {
        mealPlan.setSideDish("French Fries");
        return this;
    }

    @Override
    public MealPlanBuilder addDessert() {
        mealPlan.setDessert("Ice Cream");
        return this;
    }

    @Override
    public MealPlan getMealPlan() {
        return mealPlan;
    }
}

// Concrete Builder class for creating a Vegetarian Meal Plan.
class VegetarianMealPlanBuilder implements MealPlanBuilder {
    private MealPlan mealPlan;

    public VegetarianMealPlanBuilder() {
        this.mealPlan = new MealPlan();
    }

    @Override
    public MealPlanBuilder addMainCourse() {
        mealPlan.setMainCourse("Vegetarian Burger");
        return this;
    }

    @Override
    public MealPlanBuilder addSideDish() {
        mealPlan.setSideDish("Salad");
        return this;
    }

    @Override
    public MealPlanBuilder addDessert() {
        mealPlan.setDessert("Fruit Salad");
        return this;
    }

    @Override
    public MealPlan getMealPlan() {
        return mealPlan;
    }
}

// The Product class represents the complex object under construction, in this
// case, a Meal Plan.
class MealPlan {
    private String mainCourse;
    private String sideDish;
    private String dessert;

    public void setMainCourse(String mainCourse) {
        this.mainCourse = mainCourse;
    }

    public void setSideDish(String sideDish) {
        this.sideDish = sideDish;
    }

    public void setDessert(String dessert) {
        this.dessert = dessert;
    }

    public void listItems() {
        System.out.println("Main Course: " + mainCourse);
        System.out.println("Side Dish: " + sideDish);
        System.out.println("Dessert: " + dessert);
    }
}

// The Director class defines the order in which the construction steps
// should be called to build the final meal plan.
class MealPlanDirector {
    public MealPlan buildStandardMeal(MealPlanBuilder builder) {
        return builder.addMainCourse()
                .addSideDish()
                .addDessert()
                .getMealPlan();
    }

    public MealPlan buildVegetarianMeal(MealPlanBuilder builder) {
        return builder.addMainCourse()
                .addSideDish()
                .addDessert()
                .getMealPlan();
    }
}

// Client code to demonstrate the Meal Plan creation process.
public class Main {
    public static void main(String[] args) {
        MealPlanDirector director = new MealPlanDirector();

        System.out.println("Testing Standard Meal Plan Builder:");
        MealPlan standardMealPlan = director.buildStandardMeal(new StandardMealPlanBuilder());
        standardMealPlan.listItems();

        System.out.println("\nTesting Vegetarian Meal Plan Builder:");
        MealPlan vegetarianMealPlan = director.buildVegetarianMeal(new VegetarianMealPlanBuilder());
        vegetarianMealPlan.listItems();
    }
}
